File: proposed-projects/round-trip-experiment/reverse/20260909T184722Z_089a4294/20260910T101353Z_4b392f72/ported_implementation/src/grammar-graph/get-serialized-rule-key.ts
import { isRuleChar, isRuleCharExcluded, isRuleEnd, isRuleRef } from './type-guards';
import { RuleType } from './types';

export const KEY_TRANSLATION = {
  [RuleType.END]: 0,
  [RuleType.CHAR]: 1,
  [RuleType.CHAR_EXCLUDE]: 2,
} as const;

export const getSerializedRuleKey = (rule: unknown): string => {
  if (isRuleEnd(rule)) {
    return `${KEY_TRANSLATION[RuleType.END]}`;
  }

  if (isRuleChar(rule) || isRuleCharExcluded(rule)) {
    return `${KEY_TRANSLATION[rule.type]}-${JSON.stringify(rule.value)}`;
  }

  if (isRuleRef(rule)) {
    return `3-${rule.value}`;
  }

  throw new Error(`Unknown rule type: ${rule}`);
};

File: packages/gbnf-experiment/data/20260908T234247Z_2da5a2c0/ported_implementation/gbnf/grammar_graph/get_serialized_rule_key.py
"""Port of ``src/grammar-graph/get-serialized-rule-key.ts``."""

from __future__ import annotations

import json
from typing import Any

from .type_guards import is_rule_char, is_rule_char_excluded, is_rule_end, is_rule_ref
from .types import RuleType

KEY_TRANSLATION = {
    RuleType.END: 0,
    RuleType.CHAR: 1,
    RuleType.CHAR_EXCLUDE: 2,
}


def get_serialized_rule_key(rule: Any) -> str:
    if is_rule_end(rule):
        return f"{KEY_TRANSLATION[RuleType.END]}"

    if is_rule_char(rule) or is_rule_char_excluded(rule):
        return (
            f"{KEY_TRANSLATION[rule.type]}-"
            f"{json.dumps(rule.value, separators=(',', ':'))}"
        )
    if is_rule_ref(rule):
        return f"3-{rule.value}"
    raise ValueError(f"Unknown rule type: {rule!r}")
